File: frontend/src/lib/api/category.ts
import api from '../api'

export interface Category {
  id: number
  name: string
  slug: string
  created_at: string
}

export const categoryApi = {
  getAll: async (): Promise<Category[]> => {
    const response = await api.get('/api/v1/categories')
    return response.data
  },

  getById: async (id: number): Promise<Category> => {
    const response = await api.get(`/api/v1/categories/${id}`)
    return response.data
  },

  create: async (name: string, slug?: string): Promise<Category> => {
    const response = await api.post('/api/v1/categories', { name, slug })
    return response.data
  },

  update: async (id: number, name?: string, slug?: string): Promise<Category> => {
    const response = await api.put(`/api/v1/categories/${id}`, { name, slug })
    return response.data
  },

  delete: async (id: number): Promise<void> => {
    await api.delete(`/api/v1/categories/${id}`)
  },
}

File: backend/internal/repository/category.go
package repository

import (
	"sticky-stick/backend/internal/models"

	"gorm.io/gorm"
)

type CategoryRepository interface {
	Create(category *models.Category) error
	GetByID(id uint) (*models.Category, error)
	GetBySlug(slug string) (*models.Category, error)
	GetAll() ([]models.Category, error)
	Update(category *models.Category) error
	Delete(id uint) error
}

type categoryRepository struct {
	db *gorm.DB
}

func NewCategoryRepository(db *gorm.DB) CategoryRepository {
	return &categoryRepository{db: db}
}

func (r *categoryRepository) Create(category *models.Category) error {
	return r.db.Create(category).Error
}

func (r *categoryRepository) GetByID(id uint) (*models.Category, error) {
	var category models.Category
	err := r.db.First(&category, id).Error
	if err != nil {
		return nil, err
	}
	return &category, nil
}

func (r *categoryRepository) GetBySlug(slug string) (*models.Category, error) {
	var category models.Category
	err := r.db.Where("slug = ?", slug).First(&category).Error
	if err != nil {
		return nil, err
	}
	return &category, nil
}

func (r *categoryRepository) GetAll() ([]models.Category, error) {
	var categories []models.Category
	err := r.db.Order("name ASC").Find(&categories).Error
	return categories, err
}

func (r *categoryRepository) Update(category *models.Category) error {
	return r.db.Save(category).Error
}

func (r *categoryRepository) Delete(id uint) error {
	return r.db.Delete(&models.Category{}, id).Error
}
